File: packages/contracts/src/report.ts
import { canonical, deny, rawDigest, validate, parse } from './index.js';
import type { ValidationReport } from './generated.js';
type ObjectValue = Record<string, any>;
const RULES = ['parse', 'structure', 'integrity'];
export function validateReport(bytes: Uint8Array, profileBytes: Uint8Array,
  expectedBinding: ObjectValue): ValidationReport {
  const profile = validate('maestro.artifact.profile/1', profileBytes) as ObjectValue;
  if (bytes.length > profile.max_report_bytes) deny('REPORT_BYTES');
  const report = validate('maestro.validation.report/1', bytes) as ObjectValue;
  if (!canonical(report.binding).equals(canonical(expectedBinding)) ||
      expectedBinding.profile !== rawDigest(profileBytes) ||
      expectedBinding.configuration !== profile.configuration || expectedBinding.family !== profile.family || expectedBinding.stage !== profile.stage) deny('BINDING');
  if (!canonical(profile.required).equals(canonical(RULES)) || report.checker !== profile.validator) deny('CHECKER');
  if (report.provenance.kind !== 'fresh' || report.fresh_count !== 1 || report.reused_count !== 0) deny('PROVENANCE');
  if (report.checks.length !== 3 ||
      !canonical(report.checks.map((v: ObjectValue) => v.id)).equals(canonical(RULES))) deny('COVERAGE');
  const outcomes = report.checks.map((v: ObjectValue) => v.outcome);
  const aggregate = outcomes.includes('error') ? 'error' : outcomes.includes('fail') ? 'fail' : 'pass';
  if (report.outcome !== aggregate) deny('AGGREGATE');
  if(report.duration_ms>profile.max_duration_ms&&aggregate!=='error')deny('TIMEOUT');
  if (report.diagnostics.length > profile.max_diagnostics || report.diagnostic_count < report.diagnostics.length ||
      report.truncated !== (report.diagnostic_count > report.diagnostics.length)) deny('DIAGNOSTICS');
  const sorted = [...report.diagnostics].sort((a, b) => canonical(a).compare(canonical(b)));
  if (!canonical(sorted).equals(canonical(report.diagnostics))) deny('DIAGNOSTICS');
  if (aggregate === 'pass' && report.diagnostic_count !== 0) deny('DIAGNOSTICS');
  return report as ValidationReport;
}
export function checkFixture(input: Uint8Array, profileBytes: Uint8Array, binding: ObjectValue,
  dependencies: { id: string; bytes: Uint8Array }[] = []): ValidationReport {
  const profile = validate('maestro.artifact.profile/1', profileBytes) as ObjectValue;
  const begin = performance.now();
  const checks = RULES.map(id => ({ id, outcome: 'error' }));
  const diagnostics: ObjectValue[] = [];
  const diagnostic = (rule: string, message: string) => diagnostics.push({ rule, location: '$', message,
    hint: 'Correct the editable draft and recheck' });
  let record: ObjectValue | undefined;
  try {
    if (input.length > profile.max_input_bytes) deny('BYTES');
    parse(input);checks[0].outcome='pass';
  } catch { checks[0].outcome = 'fail'; diagnostic('STRUCTURE', 'Input rejected'); }
  if(checks[0].outcome==='pass')try{record=validate('maestro.fixture.json/1',input,profile.stage==='accepted') as ObjectValue;checks[1].outcome='pass';}catch{checks[1].outcome='fail';diagnostic('STRUCTURE','Input rejected');}
  else checks[1].outcome='fail';
  const pairs = dependencies.map(v => ({ id: v.id, digest: rawDigest(v.bytes) }));
  const unique = new Set(pairs.map(v => v.id)).size === pairs.length;
  if (record && unique && rawDigest(input) === binding.input &&
    rawDigest(canonical(pairs)) === binding.dependencies && canonical(pairs).equals(canonical(record.dependencies))) checks[2].outcome = 'pass';
  else { checks[2].outcome = 'fail'; diagnostic('INTEGRITY', 'Trusted bytes mismatch'); }
  const elapsed = Math.ceil(performance.now() - begin);
  // Preserve the historical finite report wire range; never fabricate a smaller
  // duration when the actual observation cannot be represented by that schema.
  if(elapsed>1000)deny('REPORT_DURATION');
  if (elapsed > profile.max_duration_ms) { checks[0].outcome = 'error'; diagnostic('TIMEOUT', 'Required check unavailable'); }
  diagnostics.sort((a,b) => canonical(a).compare(canonical(b)));
  const report = { schema: 'maestro.validation.report/1', binding, checker: 'maestro.fixture-checker/1',
    provenance: { kind: 'fresh' }, checks,
    outcome: checks.some(v=>v.outcome==='error') ? 'error' : checks.some(v=>v.outcome==='fail') ? 'fail' : 'pass',
    diagnostics: diagnostics.slice(0, profile.max_diagnostics), diagnostic_count: diagnostics.length,
    truncated: diagnostics.length > profile.max_diagnostics, duration_ms: elapsed,
    fresh_count: 1, reused_count: 0, repair_count: 0 };
  return validateReport(canonical(report), profileBytes, binding);
}

File: crates/maestro-contracts/src/report.rs
use crate::{Error, Result, canonical, raw_digest, validate};
use serde_json::{Value, json};
use std::{collections::HashSet, time::Instant};
const RULES: [&str; 3] = ["parse", "structure", "integrity"];
pub fn validate_report(bytes: &[u8], profile_bytes: &[u8], expected: &Value) -> Result<Value> {
    let profile = validate("maestro.artifact.profile/1", profile_bytes, false)?;
    if bytes.len() as u64
        > profile["max_report_bytes"]
            .as_u64()
            .ok_or(Error("PROFILE"))?
    {
        return Err(Error("REPORT_BYTES"));
    }
    let report = validate("maestro.validation.report/1", bytes, false)?;
    if report["binding"] != *expected
        || expected["profile"] != raw_digest(profile_bytes)
        || expected["configuration"] != profile["configuration"]
        || expected["family"] != profile["family"]
        || expected["stage"] != profile["stage"]
    {
        return Err(Error("BINDING"));
    }
    if profile["required"] != json!(RULES) || report["checker"] != profile["validator"] {
        return Err(Error("CHECKER"));
    }
    if report["provenance"] != json!({"kind":"fresh"})
        || report["fresh_count"] != 1
        || report["reused_count"] != 0
    {
        return Err(Error("PROVENANCE"));
    }
    let checks = report["checks"].as_array().ok_or(Error("COVERAGE"))?;
    if checks.len() != 3 || checks.iter().zip(RULES).any(|(v, id)| v["id"] != id) {
        return Err(Error("COVERAGE"));
    }
    let outcome = if checks.iter().any(|v| v["outcome"] == "error") {
        "error"
    } else if checks.iter().any(|v| v["outcome"] == "fail") {
        "fail"
    } else {
        "pass"
    };
    if report["outcome"] != outcome {
        return Err(Error("AGGREGATE"));
    }
    if report["duration_ms"].as_u64() > profile["max_duration_ms"].as_u64() && outcome != "error" {
        return Err(Error("TIMEOUT"));
    }
    let diagnostics = report["diagnostics"]
        .as_array()
        .ok_or(Error("DIAGNOSTICS"))?;
    let count = report["diagnostic_count"]
        .as_u64()
        .ok_or(Error("DIAGNOSTICS"))?;
    if diagnostics.len() as u64
        > profile["max_diagnostics"]
            .as_u64()
            .ok_or(Error("PROFILE"))?
        || count < diagnostics.len() as u64
        || report["truncated"] != json!(count > diagnostics.len() as u64)
        || (outcome == "pass" && count != 0)
    {
        return Err(Error("DIAGNOSTICS"));
    }
    let enc = diagnostics
        .iter()
        .map(canonical)
        .collect::<Result<Vec<_>>>()?;
    if enc.windows(2).any(|pair| pair[0] > pair[1]) {
        return Err(Error("DIAGNOSTICS"));
    }
    Ok(report)
}
pub fn check_fixture(
    input: &[u8],
    profile_bytes: &[u8],
    binding: &Value,
    dependencies: &[(String, Vec<u8>)],
) -> Result<Value> {
    let profile = validate("maestro.artifact.profile/1", profile_bytes, false)?;
    let start = Instant::now();
    let mut outcomes = ["error"; 3];
    let mut diagnostics = Vec::new();
    let diagnostic = |rule: &str, message: &str| json!({"rule":rule,"location":"$","message":message,"hint":"Correct the editable draft and recheck"});
    let record = if input.len() as u64
        <= profile["max_input_bytes"]
            .as_u64()
            .ok_or(Error("PROFILE"))?
    {
        if crate::parse(input).is_ok() {
            outcomes[0] = "pass";
            validate(
                "maestro.fixture.json/1",
                input,
                profile["stage"] == "accepted",
            )
            .ok()
        } else {
            None
        }
    } else {
        None
    };
    if record.is_some() {
        outcomes[0] = "pass";
        outcomes[1] = "pass";
    } else {
        if outcomes[0] != "pass" {
            outcomes[0] = "fail";
        }
        outcomes[1] = "fail";
        diagnostics.push(diagnostic("STRUCTURE", "Input rejected"));
    }
    let pairs = Value::Array(
        dependencies
            .iter()
            .map(|(id, bytes)| json!({"id":id,"digest":raw_digest(bytes)}))
            .collect(),
    );
    let unique = dependencies
        .iter()
        .map(|(id, _)| id)
        .collect::<HashSet<_>>()
        .len()
        == dependencies.len();
    if unique
        && record.as_ref().is_some_and(|v| v["dependencies"] == pairs)
        && binding["input"] == raw_digest(input)
        && binding["dependencies"] == raw_digest(&canonical(&pairs)?)
    {
        outcomes[2] = "pass";
    } else {
        outcomes[2] = "fail";
        diagnostics.push(diagnostic("INTEGRITY", "Trusted bytes mismatch"));
    }
    let elapsed = start.elapsed().as_millis() as u64;
    if elapsed > 1000 {
        return Err(Error("REPORT_DURATION"));
    }
    let max = profile["max_duration_ms"]
        .as_u64()
        .ok_or(Error("PROFILE"))?;
    if elapsed > max {
        outcomes[0] = "error";
        diagnostics.push(diagnostic("TIMEOUT", "Required check unavailable"));
    }
    diagnostics.sort_by_cached_key(|v| canonical(v).expect("fixed diagnostics"));
    let count = diagnostics.len();
    let limit = profile["max_diagnostics"]
        .as_u64()
        .ok_or(Error("PROFILE"))? as usize;
    diagnostics.truncate(limit);
    let report = json!({"schema":"maestro.validation.report/1","binding":binding,"checker":"maestro.fixture-checker/1","provenance":{"kind":"fresh"},"checks":RULES.iter().zip(outcomes).map(|(id,outcome)|json!({"id":id,"outcome":outcome})).collect::<Vec<_>>(),"outcome":if outcomes.contains(&"error"){"error"}else if outcomes.contains(&"fail"){"fail"}else{"pass"},"diagnostics":diagnostics,"diagnostic_count":count,"truncated":count>limit,"duration_ms":elapsed,"fresh_count":1,"reused_count":0,"repair_count":0});
    validate_report(&canonical(&report)?, profile_bytes, binding)
}
